File: app/src/main/java/android/mahendra/attendancemanager/repositories/SubjectRepository.java
package android.mahendra.attendancemanager.repositories;

import android.mahendra.attendancemanager.daos.SubjectDao;

import android.app.Application;
import android.mahendra.attendancemanager.database.MainDatabase;
import android.mahendra.attendancemanager.models.Subject;


import androidx.lifecycle.LiveData;

import java.util.List;

public class SubjectRepository {

    private SubjectDao mSubjectDao;

    private LiveData<List<Subject>> mAllSubjects;

    public SubjectRepository(Application application) {
        MainDatabase db = MainDatabase.getDatabase(application);
        mSubjectDao = db.subjectDao();
        mAllSubjects = mSubjectDao.getAllSubjects();

    }

    public LiveData<List<Subject>> getAllSubjects() {
        return mAllSubjects;
    }

    public Subject getSubject(String title) {
        for (Subject subject : mAllSubjects.getValue()) {
            if (subject.getTitle().equals(title)) {
                return subject;
            }
        }
        return null;
    }

    public void updateTitle(String oldTitle, String newTitle) {
        MainDatabase.databaseWriteExecutor.execute(() -> mSubjectDao.updateTitle(oldTitle, newTitle));
    }

    public void insert(Subject subject) {
        MainDatabase.databaseWriteExecutor.execute(() -> {
            mSubjectDao.insert(subject);
        });
    }

    public void update(Subject subject) {
        MainDatabase.databaseWriteExecutor.execute(() -> mSubjectDao.update(subject));
    }

    public void delete(Subject subject) {
        MainDatabase.databaseWriteExecutor.execute(() -> mSubjectDao.delete(subject));
    }
}

File: app/src/main/java/android/mahendra/attendancemanager/repositories/SubjectRepository.kt
package android.mahendra.attendancemanager.repositories

import android.mahendra.attendancemanager.daos.SubjectDao
import android.mahendra.attendancemanager.models.Subject
import androidx.lifecycle.LiveData

class SubjectRepository private constructor(
    private val subjectDao: SubjectDao
) {
    val allSubjects: LiveData<List<Subject>> = subjectDao.getAllSubjects()

    fun getSubject(title: String): Subject? {
        for (subject in allSubjects.value!!) {
            if (subject.title == title) {
                return subject
            }
        }
        return null
    }

    suspend fun updateTitle(oldTitle: String?, newTitle: String?) {
        subjectDao.updateTitle(oldTitle!!, newTitle!!)
    }

    suspend fun insert(subject: Subject?) {
        subjectDao.insert(subject!!)
    }

    suspend fun update(subject: Subject?) {
        subjectDao.update(subject!!)
    }

    suspend fun incrementAttendedClasses(title: String) {
        subjectDao.incrementAttendedClasses(title)
    }

    suspend fun incrementMissedClasses(title: String) {
        subjectDao.incrementMissedClasses(title)
    }

    suspend fun delete(subject: Subject?) {
        subjectDao.delete(subject!!)
    }

    suspend fun updateAttendance(title: String, attendedClasses: Int, missedClasses: Int) {
        subjectDao.updateAttendance(title, attendedClasses, missedClasses)
    }

    suspend fun resetAttendance(title: String) {
        updateAttendance(title, 0, 0)
    }

    companion object {

        @Volatile private var instance: SubjectRepository? = null

        fun getInstance(subjectDao: SubjectDao) =
                instance ?: synchronized(this) {
                    instance ?: SubjectRepository(subjectDao).also { instance = it }
                }
    }
}